clippy_lints/src/mutable_debug_assertion.rs

```diff
@@ -0,0 +1,105 @@
+use crate::utils::{is_direct_expn_of, span_lint};
+use if_chain::if_chain;
+use rustc::hir::{Expr, ExprKind, Mutability, StmtKind, UnOp};
+use rustc::lint::{LateContext, LateLintPass, LintArray, LintPass};
+use rustc::{declare_lint_pass, declare_tool_lint};
+use syntax_pos::Span;
+
+declare_clippy_lint! {
+    /// **What it does:** Checks for function/method calls with a mutable
+    /// parameter in `debug_assert!`, `debug_assert_eq!` and `debug_assert_ne!` macros.
+    ///
+    /// **Why is this bad?** In release builds `debug_assert!` macros are optimized out by the
+    /// compiler.
+    /// Therefore mutating something in a `debug_assert!` macro results in different behaviour
+    /// between a release and debug build.
+    ///
+    /// **Known problems:** None
+    ///
+    /// **Example:**
+    /// ```rust,ignore
+    /// debug_assert_eq!(vec![3].pop(), Some(3));
+    /// // or
+    /// fn take_a_mut_parameter(_: &mut u32) -> bool { unimplemented!() }
+    /// debug_assert!(take_a_mut_parameter(&mut 5));
+    /// ```
+    pub DEBUG_ASSERT_WITH_MUT_CALL,
+    correctness,
+    "`debug_assert!` and others will be ignored in release mode and therefore alter the execution flow. Replace it with non-mut function calls."
+}
+
+declare_lint_pass!(DebugAssertWithMutCall => [DEBUG_ASSERT_WITH_MUT_CALL]);
+
+fn check_for_mutable_call(cx: &LateContext<'_, '_>, e: &Expr) -> bool {
+    match &e.kind {
+        ExprKind::AddrOf(Mutability::MutMutable, _) => return true,
+        ExprKind::Call(_, args) => return (*args).iter().any(|a| check_for_mutable_call(cx, a)),
+        /* TODO: I can't detect `&mut self`. How to? */
+        ExprKind::MethodCall(_, _, args) => return (*args).iter().any(|a| check_for_mutable_call(cx, a)),
```

Comment:
I need some help here. The `expr_ty` doesn't work for me sadyly, because it returns an [`TyKind::Adt`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/ty/enum.TyKind.html#variant.Adt) for the first argument. I'm really not sure how to differentiate between `self`, `&self` and `&mut self`.
If you don't know an answer, can you ping someone who might have a clue?

Reply:
Yeah, just tried that myself for a while, but didn't find a solution.

@rust-lang/clippy any idea on how to find out if `self` is passed as `&mut` on a `MethodCall`?

Reply:
perhaps checking the "adjustment type" will get us that?

Reply:
Awesome!

> Now I only have strugle with `*mut T`

Does matching on [`ty::RawPtr`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/ty/enum.TyKind.html#variant.RawPtr) not work?

Reply:
Sadly not :/ Not sure why though. Not sure, if it's really that relevant ^^